### PR TITLE
feat: program counter based gas used counter

### PR DESCRIPTION
**Pull request by Will-Smith11**
https://sorella-labs.atlassian.net/jira/software/projects/AG/boards/3?selectedIssue=AG-206

This is the underlying simulation logic. want to get this merged in before i finish #232  and then ill add some more tests and cleanup there.

### CODE REVIEW

File: crates/consensus/src/contract/mod.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    // If they're in opposite directions then we see if we have to flip them
+                    if tob.2 > amm.3 {
+                        Some((tob.0, tob.1, tob.2 - amm.2, tob.3 - amm.3))
+                    } else {
+                        Some((amm.0, amm.1, amm.2 - tob.3, amm.3 - tob.2))
+                    }
+                }
+            }
+            (net_amm_order, tob_swap) => net_amm_order.or(tob_swap)
+        };
```
Comment (jnoorchashm37) [inline crates/consensus/src/contract/mod.rs:94]: maybe a function? kinda ugly

File: crates/consensus/src/contract/mod.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            asset_out_index as usize,
+            quantity_in,
+            quantity_out
+        );
+        // Account for our reward
+        asset_builder.allocate(AssetBuilderStage::Reward, t0, tob_outcome.total_reward.to());
+        let rewards_update = tob_outcome.to_rewards_update();
+        // Push the pool update
+        pool_updates.push(PoolUpdate {
+            zero_for_one: false,
```
Comment (jnoorchashm37) [inline crates/consensus/src/contract/mod.rs:114]: doesn't this have b calculated by checked the asset indexes?

File: crates/types/src/sol_bindings/ext/grouped_orders.rs [hunk 3]
```diff
@@ -559,7 +560,12 @@ impl RawPoolOrder for PartialStandingOrder {
     }
 
     fn amount_out_min(&self) -> u128 {
-        self.amountFilled
+        // TODO: verify math on this. feels wrong
+        if self.assetIn < self.assetOut {
+            self.minAmountIn * self.minPrice.to::<u128>()
+        } else {
+            self.minAmountIn / self.minPrice.to::<u128>()
+        }
```
Comment (jnoorchashm37) [inline crates/types/src/sol_bindings/ext/grouped_orders.rs:568]: i don't think this is correct either.

File: crates/consensus/src/contract/mod.rs (hunk 2 above)
Comment (Will-Smith11) [reply crates/consensus/src/contract/mod.rs:114]: not my code

**Pull request merged**